File: src/main/java/pl/training/workshop/m5/s12_bridgemethods/start/RuleRegistry.java
package pl.training.workshop.m5.s12_bridgemethods.start;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import pl.training.workshop.shared.Money;
import pl.training.workshop.m5.s12_bridgemethods.Ticket;

/**
 * Start: rejestr "frameworkowy" - dla każdej metody o nazwie apply bierze typ parametru jako klucz.
 * Działa, dopóki w klasie reguły jest dokładnie jedna metoda apply.
 */
public final class RuleRegistry {
    private final Map<Class<?>, Object> rules = new HashMap<>();

    public RuleRegistry(Object... rules) {
        for (Object rule : rules) {
            for (Method method : rule.getClass().getDeclaredMethods()) {
                if (method.getName().equals("apply")) {
                    this.rules.put(method.getParameterTypes()[0], rule);
                }
            }
        }
    }

    public Money price(Ticket ticket) {
        Object rule = rules.get(ticket.getClass());
        try {
            return (Money) rule.getClass().getMethod("apply", ticket.getClass()).invoke(rule, ticket);
        } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException e) {
            throw new IllegalStateException("brak reguły dla " + ticket.getClass().getSimpleName(), e);
        }
    }

    public List<String> supportedTypes() {
        return rules.keySet().stream().map(Class::getSimpleName).sorted().toList();
    }

    public static RuleRegistry standard() {
        return new RuleRegistry(new StandardRule(), new StudentRule());
    }
}

File: csharp/src/Training.Workshop/M5/S12BridgeMethods/Start/RuleRegistry.cs
using System.Reflection;
using Training.Workshop.Shared;

namespace Training.Workshop.M5.S12BridgeMethods.Start;

/// <summary>
/// Start: rejestr "frameworkowy" - dla każdej metody o nazwie Apply bierze typ parametru jako klucz.
/// Działa, dopóki w klasie reguły jest dokładnie jedna metoda Apply.
/// </summary>
public sealed class RuleRegistry
{
    private readonly Dictionary<Type, object> _rules = [];

    public RuleRegistry(params object[] rules)
    {
        foreach (var rule in rules)
        {
            var type = rule.GetType();
            foreach (var method in type.GetMethods(BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.DeclaredOnly))
            {
                if (method.Name == "Apply")
                {
                    _rules[method.GetParameters()[0].ParameterType] = rule;
                }
            }
        }
    }

    public Money Price(ITicket ticket)
    {
        var rule = _rules.GetValueOrDefault(ticket.GetType());
        var method = rule!.GetType().GetMethod("Apply", [ticket.GetType()])
            ?? throw new InvalidOperationException("brak reguły dla " + ticket.GetType().Name);
        try
        {
            return (Money)method.Invoke(rule, [ticket])!;
        }
        catch (TargetInvocationException e)
        {
            throw new InvalidOperationException("brak reguły dla " + ticket.GetType().Name, e);
        }
    }

    public IReadOnlyList<string> SupportedTypes()
    {
        return _rules.Keys.Select(t => t.Name).Order(StringComparer.Ordinal).ToList();
    }

    public static RuleRegistry Standard()
    {
        return new RuleRegistry(new StandardRule(), new StudentRule());
    }
}
